kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java

```diff
@@ -402,19 +412,34 @@ void cacheRealmPermsChanged(String realmRef, Set<String> roles, Set<String> perm
             if (cacheDebug) M_log.info("SScache:changed .anon:found in "+azgRef);
             for (String perm : permissions) {
                 if (perm != null) {
-                    keysToInvalidate.add(makeCacheKey(null, perm, azgRef, false));
+                    keysToInvalidate.add(makeCacheKey(null, perm, azgRef, false, false));
                 }
             }
         }
         // now handle all the real users
+	// clear both normal and swapped users
+	// start with a set of all roles to which one can swap
+	Set<String> svRolesFinal = (Set<String>)svRoles.clone();
+	svRolesFinal.retainAll(roles);  
+
         Set<Member> members = azg.getMembers();
         if (members != null && !members.isEmpty()) {
             for (Member member : members) {
                 if (member != null && member.isActive() && member.getUserId() != null) {
+		    boolean canSwap = (member.getRole().isAllowed(SiteService.SITE_ROLE_SWAP));
                     for (String perm : permissions) {
                         if (perm != null) {
-                            keysToInvalidate.add(makeCacheKey(member.getUserId(), perm, azgRef, false));
-                        }
+                            keysToInvalidate.add(makeCacheKey(member.getUserId(), perm, azgRef, false, false));
+			    // Only invalidate swapped roles if the user can swap
+			    // This is an approximation. If a user is swapped and their permission to swap is removed
+			    // or the role they are swapped to has been removed from the site
+			    // we will not invalidate their data. Their info may wait until the expiration time to sync up
+			    if (canSwap) {
+				for (String invRole: svRolesFinal) {
+				    keysToInvalidate.add(makeCacheKey(invRole + "!" + member.getUserId(), perm, azgRef, false, false));
```

Comment:
I don't think that makes it simpler. It's just @ rather than !. I was concerned with whether other parts of the code depends upon the format of the key. If I can't prove that no one cares, it's safer not to change it.